sky/packages/sky/lib/painting/paragraph_painter.dart

```diff
@@ -0,0 +1,147 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:sky' as sky;
+
+import 'package:sky/painting/text_style.dart';
+
+// This must be immutable, because we won't notice when it changes
+abstract class TextSpan {
+  sky.Node _toDOM(sky.Document owner);
+  String toString([String prefix = '']);
+
+  void _applyStyleToContainer(sky.Element container) {
+  }
+}
+
+class PlainTextSpan extends TextSpan {
+  PlainTextSpan(this.text) {
+    assert(text != null);
+  }
+
+  final String text;
+
+  sky.Node _toDOM(sky.Document owner) {
+    return owner.createText(text);
+  }
+
+  bool operator ==(other) => other is PlainTextSpan && text == other.text;
+  int get hashCode => text.hashCode;
+
+  String toString([String prefix = '']) => '${prefix}InlineText: "${text}"';
+}
+
+class StyledTextSpan extends TextSpan {
+  StyledTextSpan(this.style, this.children) {
+    assert(style != null);
+    assert(children != null);
+  }
+
+  final TextStyle style;
+  final List<TextSpan> children;
+
+  sky.Node _toDOM(sky.Document owner) {
+    sky.Element parent = owner.createElement('t');
+    style.applyToCSSStyle(parent.style);
+    for (TextSpan child in children) {
+      parent.appendChild(child._toDOM(owner));
+    }
+    return parent;
+  }
+
+  void _applyStyleToContainer(sky.Element container) {
+    style.applyToContainerCSSStyle(container.style);
+  }
+
+  bool operator ==(other) {
+    if (identical(this, other))
+      return true;
+    if (other is! StyledTextSpan
+        || style != other.style
+        || children.length != other.children.length)
+      return false;
+    for (int i = 0; i < children.length; ++i) {
+      if (children[i] != other.children[i])
+        return false;
+    }
+    return true;
+  }
+
+  int get hashCode {
```

Comment:
Quiver can do this for us.  I believe we (at least transitively) depend on quiver these days.